src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs

```diff
@@ -27,126 +28,210 @@ public partial class Ping
+            if (!_ipv6)
             {
-                UnregisterWaitHandle();
-                throw;
+                return (int)Interop.IpHlpApi.IcmpSendEcho2(
+                    _handlePingV4,
+                    GetWaitHandle(isAsync),
+                    IntPtr.Zero,
+                    IntPtr.Zero,
+#pragma warning disable CS0618 // Address is marked obsolete
+                        (uint)address.Address,
+#pragma warning restore CS0618
+                        _requestBuffer,
+                    (ushort)buffer.Length,
+                    ref ipOptions,
+                    _replyBuffer,
+                    MaxUdpPacket,
+                    (uint)timeout);
             }
 
-            if (error == 0)
+            IPEndPoint ep = new IPEndPoint(address, 0);
+            Internals.SocketAddress remoteAddr = IPEndPointExtensions.Serialize(ep);
+            byte[] sourceAddr = new byte[28];
+
+            return (int)Interop.IpHlpApi.Icmp6SendEcho2(
+                _handlePingV6,
+                GetWaitHandle(isAsync),
+                IntPtr.Zero,
+                IntPtr.Zero,
+                sourceAddr,
+                remoteAddr.Buffer,
+                _requestBuffer,
+                (ushort)buffer.Length,
+                ref ipOptions,
+                _replyBuffer,
+                MaxUdpPacket,
+                (uint)timeout);
+        }
+
+        private PingReply CreatePingReply()
+        {
+            SafeLocalAllocHandle buffer = _replyBuffer;
+
+            // Marshals and constructs new reply.
+            if (_ipv6)
             {
-                error = Marshal.GetLastWin32Error();
+                Interop.IpHlpApi.Icmp6EchoReply icmp6Reply = Marshal.PtrToStructure<Interop.IpHlpApi.Icmp6EchoReply>(buffer.DangerousGetHandle());
+                return CreatePingReplyFromIcmp6EchoReply(icmp6Reply, buffer.DangerousGetHandle(), _sendSize);
+            }
 
-                // Only skip Async IO Pending error value.
-                if (error != Interop.IpHlpApi.ERROR_IO_PENDING)
-                {
-                    // Cleanup.
-                    FreeUnmanagedStructures();
-                    UnregisterWaitHandle();
+            Interop.IpHlpApi.IcmpEchoReply icmpReply = Marshal.PtrToStructure<Interop.IpHlpApi.IcmpEchoReply>(buffer.DangerousGetHandle());
+            return CreatePingReplyFromIcmpEchoReply(icmpReply);
+        }
 
-                    throw new Win32Exception(error);
-                }
+        private void Cleanup(bool isAsync, bool isComplete)
+        {
+            FreeUnmanagedStructures();
+
+            if (isAsync)
+            {
+                UnregisterWaitHandle();
             }
 
-            return tcs.Task;
+            if (isComplete)
+            {
+                Finish();
+            }
         }
 
-        /*private*/ partial void InternalDisposeCore()
+        /*private*/
+        partial void InternalDisposeCore()
```

Comment:
What's going on here? Is this supposed to be private or not?

Reply:
This is just the way it was previously but I can remove /*private*/